src/preprocessing/passes/bool_to_bv.cpp

```diff
@@ -39,165 +39,159 @@ PreprocessingPassResult BoolToBV::applyInternal(
       options::preprocessStep());
 
   unsigned size = assertionsToPreprocess->size();
-  for (unsigned i = 0; i < size; ++i)
+
+  if (options::boolToBitvector() == BOOL_TO_BV_ALL)
+  {
+    for (unsigned i = 0; i < size; ++i)
+    {
+      Node newAssertion = lowerAssertion((*assertionsToPreprocess)[i], true);
+      assertionsToPreprocess->replace(i, Rewriter::rewrite(newAssertion));
+    }
+  }
+  else if (options::boolToBitvector() == BOOL_TO_BV_ITE)
+  {
+    for (unsigned i = 0; i < size; ++i)
+    {
+      assertionsToPreprocess->replace(
+          i, Rewriter::rewrite(lowerIte((*assertionsToPreprocess)[i])));
+    }
+  }
+  else
   {
-    assertionsToPreprocess->replace(
-        i, Rewriter::rewrite(lowerAssertion((*assertionsToPreprocess)[i])));
+    // Unhandled option
+    Unreachable();
   }
 
   return PreprocessingPassResult::NO_CONFLICT;
 }
 
+void BoolToBV::updateCache(TNode n, TNode rebuilt)
+{
+  // check more likely case first
+  if ((n.getKind() != kind::ITE) || !n[1].getType().isBitVector())
+  {
+    d_lowerCache[n] = rebuilt;
+  }
+  else
+  {
+    d_iteBVLowerCache[n] = rebuilt;
+  }
+}
+
 Node BoolToBV::fromCache(TNode n) const
 {
-  if (d_lowerCache.find(n) != d_lowerCache.end())
+  // check more likely case first
+  if (n.getKind() != kind::ITE)
+  {
+    if (d_lowerCache.find(n) != d_lowerCache.end())
+    {
+      return d_lowerCache.at(n);
+    }
+  }
+  else
   {
-    return d_lowerCache.find(n)->second;
+    if (d_iteBVLowerCache.find(n) != d_iteBVLowerCache.end())
+    {
+      return d_iteBVLowerCache.at(n);
+    }
   }
   return n;
 }
 
+inline bool BoolToBV::inCache(const Node& n) const
+{
+  return (ContainsKey(d_lowerCache, n) || ContainsKey(d_iteBVLowerCache, n));
+}
+
 bool BoolToBV::needToRebuild(TNode n) const
 {
   // check if any children were rebuilt
   for (const Node& nn : n)
   {
-    if (ContainsKey(d_lowerCache, nn))
+    if (inCache(nn))
     {
       return true;
     }
   }
   return false;
 }
 
-Node BoolToBV::lowerAssertion(const TNode& a)
+Node BoolToBV::lowerAssertion(const TNode& assertion, bool force)
+{
+  // first try to lower all the children
+  for (const Node& c : assertion)
+  {
+    lowerNode(c, force);
+  }
+
+  // now try lowering the assertion, but don't force it (even in mode all)
+  lowerNode(assertion, false);
+  Node newAssertion = fromCache(assertion);
+  TypeNode newAssertionType = newAssertion.getType();
+  if (newAssertionType.isBitVector())
+  {
+    Assert(newAssertionType.getBitVectorSize() == 1);
+    NodeManager* nm = NodeManager::currentNM();
+    newAssertion = nm->mkNode(kind::EQUAL, newAssertion, bv::utils::mkOne(1));
```

Comment:
You could merge these two lines since you use `nm` only once.

Reply:
Done with 7367bdb1ad9a3e59afedfff86ad8df441c6c6794